engine/execution/pruner/core.go

```diff
@@ -0,0 +1,176 @@
+			// also we can disable this feature by setting the sleep time to a very large value.
+			// also allows the pruner to be more responsive to the context cancellation, meaning
+			// while the pruner is sleeping, it can be cancelled immediately.
+		case <-time.After(config.SleepAfterEachIteration):
+		}
+
+		iter, hasNext, err := creator.Create()
+		if err != nil {
+			return fmt.Errorf("failed to create block iterator: %w", err)
+		}
+
+		if !hasNext {
+			// no more blocks to iterate, we are done.
+			continue
+		}
+
+		err = iterateAndPruneAll(iter)
+		if err != nil {
+			return fmt.Errorf("failed to iterate, execute, and commit in batch: %w", err)
+		}
+	}
+}
+
+// makeBlockIteratorCreator create the block iterator creator
+func makeBlockIteratorCreator(
+	state protocol.State,
+	badgerDB *badger.DB,
+	headers storage.Headers,
+	chunkDataPacksDB storage.DB,
+	config PruningConfig,
+) (
+	module.IteratorCreator,
+	// this is for logging purpose, so that after each round of pruning,
+	// we can log and report metrics about the next and latest to prune
+	func() (nextToPrune uint64, latestToPrune uint64, err error),
+	error, // any error are exception
+) {
+	root := state.Params().SealedRoot()
+	sealedAndExecuted := latest.NewLatestSealedAndExecuted(
+		root,
+		state,
+		badgerDB,
+	)
+
+	// retrieves the latest sealed and executed block height.
+	// the threshold ensures that a certain number of blocks are retained for querying instead of being pruned.
+	latest := &LatestPrunable{
+		LatestSealedAndExecuted: sealedAndExecuted,
+		threshold:               config.Threshold,
+	}
+
+	initializer := store.NewConsumerProgress(chunkDataPacksDB, NextHeightForUnprunedExecutionDataPackKey)
+
+	creator, err := block_iterator.NewHeightBasedCreator(
+		headers.BlockIDByHeight,
+		initializer,
+		root,
+		latest.Latest,
+	)
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create height based block iterator creator: %w", err)
+	}
+
+	stateReader := creator.IteratorState()
+
+	return creator, func() (nextToPrune uint64, latestToPrune uint64, err error) {
+		next, err := stateReader.LoadState()
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to get next height to prune: %w", err)
+		}
+
+		header, err := latest.Latest()
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to get latest prunable block: %w", err)
+		}
+
+		return next, header.Height, nil
+	}, nil
+}
+
+// makeIterateAndPruneAll takes config and chunk data packs db and pruner and returns a function that
+// takes a block iterator and iterates through all the blocks and decides how to prune the chunk data packs.
+func makeIterateAndPruneAll(
+	log zerolog.Logger, ctx context.Context, config PruningConfig, chunkDataPacksDB storage.DB, prune *ChunkDataPackPruner,
+) func(iter module.BlockIterator) error {
+	isBatchFull := func(counter int) bool {
+		return uint(counter) >= config.BatchSize
+	}
+
+	sleeper := func() {
+		time.Sleep(config.SleepAfterEachBatchCommit)
```

Comment:
this function also needs to pay attention to the context.

Reply:
It might be better to just put this logic inside `IterateExecuteAndCommitInBatch`

Reply:
I changed the sleeper function to listen to the context.Done(). 

>It might be better to just put this logic inside IterateExecuteAndCommitInBatch

I can't, because when `time.Sleep()` starts, it can't be stopped, unless it's in a `select` statement. So I had to put the select context.Done() into this sleeper function.